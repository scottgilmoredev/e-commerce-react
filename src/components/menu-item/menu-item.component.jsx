import React from 'react';
import { withRouter } from 'react-router-dom';

// Styles
import {
    BackgroundImageContainer,
    ContentContainer,
    ContentSubtitle,
    ContentTitle,
    MenuItemContainer,
} from './menu-item.styles';

/**
 * Display a menu item.
 * Menu items contain an image, a title, and a sub-title.
 * @param {String} imageUrl - the url for the image.
 * @param {String} linkUrl - the url to link to.
 * @param {String} size - menu item size.
 * @param {String} title - menu item title.
 * @param {Object} history - react router dom history.
 * @param {Object} match - react router dom match.
 */
export const MenuItem = ({ imageUrl, linkUrl, size, title, history, match }) => (
    <MenuItemContainer
        size={ size }
        onClick={() => history.push(`${ match.url }${ linkUrl }`)}    
    >
        { /* Menu item image */ }
        <BackgroundImageContainer
            className='background-image'
            imageUrl={ imageUrl }
        />

        { /* Menu item content */ }
        <ContentContainer className='content'>
            <ContentTitle>{ title.toUpperCase() }</ContentTitle>
            <ContentSubtitle>SHOP NOW</ContentSubtitle>
        </ContentContainer>
    </MenuItemContainer>
);

export default withRouter(MenuItem);
