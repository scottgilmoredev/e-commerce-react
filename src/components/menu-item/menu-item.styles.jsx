import styled from 'styled-components';

export const MenuItemContainer = styled.div`
    align-items: center;
    border: 1px solid black;
    display: flex;
    flex: 1 1 auto;
    height: ${({ size }) => (size ? '380px' : '240px')};
    justify-content: center;
    margin: 0 7.5px 15px;
    min-width: 30%;
    overflow: hidden;

    &:hover {
        cursor: pointer;
        
        & .background-image {
            transform: scale(1.1);
            transition: transform 6s cubic-bezier(0.25, 0.45, 0.45, 0.95);
        }

        & .content {
            opacity: 0.9;
        }
    }

    &:first-child {
        margin-right: 7.5px;
    }

    &:last-child {
        margin-left: 7.5px;
    }

    @media screen and (max-width: 800px) {
        height: 200px;

        &.large {
            height: 200px;
        }
    }
`;

export const BackgroundImageContainer = styled.div`
    background-image: ${({ imageUrl }) => `url(${imageUrl})`};
    background-position: center;
    background-size: cover;
    height: 100%;
    width: 100%;
`;

export const ContentContainer = styled.div`
    align-items: center;
    background-color: white;
    border: 1px solid black;
    display: flex;
    flex-direction: column;
    height: 90px;
    justify-content: center;
    opacity: 0.7;
    padding: 0 25px;
    position: absolute;
`;

export const ContentTitle = styled.span`
    color: #4a4a4a;
    font-size: 22px;
    font-weight: bold;
    margin-bottom: 6px;
`;

export const ContentSubtitle = styled.span`
    font-size: 16px;
    font-weight: lighter;
`;

// Display names.
MenuItemContainer.displayName = 'MenuItemContainer';
BackgroundImageContainer.displayName = 'BackgroundImageContainer';
ContentContainer.displayName = 'ContentContainer';
ContentTitle.displayName = 'ContentTitle';
ContentSubtitle.displayName = 'ContentSubtitle';
