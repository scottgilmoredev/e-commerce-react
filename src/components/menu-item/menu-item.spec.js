import React from 'react';
import { shallow } from 'enzyme';

// Component
import { MenuItem } from './menu-item.component';

describe('MenuItem component', () => {
    let wrapper;
    let mockMatch;
    let mockHistory;
    const linkUrl = '/hats';
    const size = 'large';
    const imageUrl = 'testImage';

    beforeEach(() => {
        mockMatch = {
            url: '/shop',
        };

        mockHistory = {
            push: jest.fn(),
        };

        const mockProps = {
            match: mockMatch,
            history: mockHistory,
            linkUrl,
            size,
            title: 'hats',
            imageUrl,
        };

        wrapper = shallow(<MenuItem { ...mockProps } />);
    });

    // Render the component.
    it('should render the MenuItem component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    // Navigate to correct route.
    it('should  call history.push with the correct string when MenuItemContainer is clicked', () => {
        wrapper.find('MenuItemContainer').simulate('click');
        expect(mockHistory.push).toHaveBeenCalledWith(`${ mockMatch.url }${ linkUrl }`);
    });

    // Pass size as prop.
    it('should pass size to MenuItemContainer as the prop size', () => {
        expect(wrapper.find('MenuItemContainer').prop('size')).toBe(size);
    });

    // Pass image url as prop.
    it('should pass imageUrl to BackgroundImageContainer as the prop imageUrl', () => {
        expect(wrapper.find('BackgroundImageContainer').prop('imageUrl')).toBe(imageUrl);
    });
});
