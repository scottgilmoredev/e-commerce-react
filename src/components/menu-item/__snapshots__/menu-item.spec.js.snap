// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem component should render the MenuItem component 1`] = `ShallowWrapper {}`;
